framework/src/meshgenerators/MeshGenerator.C

```diff
@@ -20,15 +20,38 @@ MeshGenerator::validParams()
                         false,
                         "Whether or not to show mesh info after generating the mesh "
                         "(bounding box, element types, sidesets, nodesets, subdomains, etc)");
-
+  params.addParam<std::vector<std::string>>(
+      "selected_mesh_metadata_to_retain",
+      std::vector<std::string>(),
+      "Names of input mesh's metadata that are selected to retain.");
+  params.addParam<bool>("retain_all_input_mesh_metadata",
+                        false,
+                        "Whether to retain all the mesh metadata of the input mesh.");
   params.registerBase("MeshGenerator");
 
   return params;
 }
 
 MeshGenerator::MeshGenerator(const InputParameters & parameters)
-  : MooseObject(parameters), MeshMetaDataInterface(this), _mesh(_app.actionWarehouse().mesh())
+  : MooseObject(parameters),
+    MeshMetaDataInterface(this),
+    _mesh(_app.actionWarehouse().mesh()),
+    _selected_mesh_metadata_to_retain(
+        getParam<std::vector<std::string>>("selected_mesh_metadata_to_retain")),
+    _retain_all_input_mesh_metadata(getParam<bool>("retain_all_input_mesh_metadata"))
 {
+  checkMeshMetadataForwardingSetting(this,
+                                     isParamValid("input"),
+                                     _retain_all_input_mesh_metadata,
+                                     _selected_mesh_metadata_to_retain);
+  if (_retain_all_input_mesh_metadata)
+    declareAllForwardedMeshMetadata(
+        getParam<MeshGeneratorName>("input"), _forwarded_metadata_names, _forwarded_metadata_types);
+  else if (!_selected_mesh_metadata_to_retain.empty())
+    declareSelectedForwardedMeshMetadata(getParam<MeshGeneratorName>("input"),
+                                         _selected_mesh_metadata_to_retain,
+                                         _forwarded_metadata_names,
+                                         _forwarded_metadata_types);
```

Comment:
this still runs before most metadata is declared by the MG.

The MG base constructor is called before meta data declaration

Reply:
The `input` MG constructor must be run before the MG constructor that needs to copy `input` mesh's metadata. As metadata can only be declared in constructor, we should be safe here.